### DISCUSSION TITLE
v1.38.0

### DISCUSSION THREAD

**Discussion by mikecao** [umami-software/umami#1486] [General]
## News
We have made lots of updates to how database migrations work. Hopefully this will provide a much smoother process going forward. We've moved the `check-db` script which performs migrations to the build step.  Now when you deploy to a host like Vercel or Netlify, your database will be upgrade before the app ever starts. This should cut down on a lot of the migration related bugs people have been running into.

For Docker users, the `check-db` can't be run during build because the images are pre-built. It will run before starting the application just like before.

We have also made major database changes to get ready for our hosted cloud offering, adding support for Clickhouse, Kafka and Redis. Any changes we make will of course be open-sourced so you will get the same performance benefits we gain. If you are interested in joining our free beta program before we launch, you can sign up at https://umami.is/pricing.

## Features
- Added `Yesterday` option to the date picker
- You can now default the theme and locale using query parameters.
  - For the theme you can use `?theme=dark` or `?theme=light`.
  - For locale you can use `?locale=en-US`.
- Admins can now reassign websites to other users
- Removed cascading deletes from the database. We now run all commands in a transaction.

## Fixes
- Fixed issue with `Edit Dashboard` not working when a website is deleted
- Fixed deployment issue with Netlify
- Fixed country lookup returning Unknown in some cases
- Fixed issue with `+` sign disappearing in the metrics display

## Updates
- Updated Next to `12.2.5`
- Updated Prisma to `4.3.1`
- Language translation updates

A huge thanks to all the contributors on this release! @zx900930 @the-lazy-fox @guigrpa @aidanm1999 @wangyang0210 @Dhravya @Maxime-J @briancao @franciscao633 @AkashRajpurohit

<hr /><em>This discussion was created from the release <a href='https://github.com/umami-software/umami/releases/tag/v1.38.0'>v1.38.0</a>.</em>

**Comment by marc-on-github** [comment]
Thanks for that update. I'm on vercel + supabase and I got this error:
![image](https://user-images.githubusercontent.com/7252726/188589297-9c25d036-814c-4451-8f7a-d858f2865e37.png)

**Reply by rohandebsarkar** [reply]
Run the following code on your supabase project SQL editor.
```
ALTER TABLE account OWNER TO postgres;
ALTER TABLE website OWNER TO postgres;
ALTER TABLE session OWNER TO postgres;
ALTER TABLE pageview OWNER TO postgres;
ALTER TABLE event OWNER TO postgres;
ALTER TABLE event_data OWNER TO postgres;
```

**Reply by marc-on-github** [reply]
Nice, thanks, looks a bit better, but now I'm getting this error: 
![image](https://user-images.githubusercontent.com/7252726/188593497-8cb006ce-0de4-4dbc-a055-adfc1c0247ce.png)

**Reply by rohandebsarkar** [reply]
Ah!, you would have to run this command locally on your system after creating the `DATABASE_URL` environment variable for manually resolving the prisma migration:
```
yarn prisma migrate resolve --applied "02_add_event_data"
```
see #1491.

**Reply by marc-on-github** [reply]
> Ah!, you can either run this command locally on your system after creating the `DATABASE_URL` environment variable for manually resolving the prisma migration:
> 
> ```
> yarn prisma migrate resolve --applied "02_add_event_data"
> ```
> 
> see #1491.

That works, perfect, thank you!

**Reply by dyxang** [reply]
I'm same issue on vercel + supabase,updated from 1.37.0 to 1.38.0.
`Error:P3009`

How to run this command on vercel?
```
yarn prisma migrate resolve --applied "02_add_event_data"
```
Is it in this position? 
I'm very sorry that my native language is not English. I may have misunderstood it.
![image](https://user-images.githubusercontent.com/36600710/188622993-60833025-ecff-4dd4-9c8c-504b2a03a148.png)

**Reply by marc-on-github** [reply]
I have run it locally.

**Reply by rohandebsarkar** [reply]
> I'm same issue on vercel + supabase,updated from 1.37.0 to 1.38.0.
> `Error:P3009`
> 
> How to run this command on vercel?
> ```
> yarn prisma migrate resolve --applied "02_add_event_data"
> ```
> Is it in this position? 
> I'm very sorry that my native language is not English. I may have misunderstood it.
> ![image](https://user-images.githubusercontent.com/36600710/188622993-60833025-ecff-4dd4-9c8c-504b2a03a148.png)
> 
> 

Just clone the repo locally in your system, declare the `DATABASE_URL` evironment variable. And run the following commands:
```
yarn install
yarn build
yarn prisma migrate resolve --applied "02_add_event_data"
```

**Reply by dyxang** [reply]
Thanks a lot.It works.

**Reply by candideu** [reply]
I ran into the same issue, and wanted to thank you all for your insight and help.

In case anyone reading this needs further clarification, here's what I did:

1. Forked the [umami](https://github.com/umami-software/umami) GitHub project
2. Cloned my fork locally using GitHub Desktop
3. In my local folder, I created a file called `.env`. In this file, I typed in `DATABASE_URL=` and pasted the postgres URL from Supabase (`Connection Pooling` > `Connection String`) along with `?pgbouncer=true`
4. Once that was done, I ran the following commands in my command prompt (with the directory set to my local cloned umami fork):
```
yarn install
yarn build
yarn prisma migrate resolve --applied "02_add_event_data"
```

**Reply by ever-blogger** [reply]
![image](https://user-images.githubusercontent.com/63139255/198755034-9ef842a9-8fac-4f84-9030-d953754bf0cd.png)
I am following the step you given, error P3009 return after `yarn build` , this is the error I always come across it's stop me to build via Vercel.

**Reply by candideu** [reply]
@ever-blogger If you're using Supabase, try [this query](https://github.com/supabase/supabase/discussions/9314#discussioncomment-4082332):

```sql
grant all privileges on all tables in schema public to postgres, anon, authenticated, service_role;
grant all privileges on all functions in schema public to postgres, anon, authenticated, service_role;
grant all privileges on all sequences in schema public to postgres, anon, authenticated, service_role;

alter default privileges in schema public grant all on tables to postgres, anon, authenticated, service_role;
alter default privileges in schema public grant all on functions to postgres, anon, authenticated, service_role;
alter default privileges in schema public grant all on sequences to postgres, anon, authenticated, service_role;
```

Then run `yarn prisma migrate resolve --applied "02_add_event_data"` again in your command prompt.

Disclaimer: I'm not a sql expert, so I'd consult the [original thread](https://github.com/supabase/supabase/discussions/9314#discussioncomment-4082332) and use at your own risk.

**Comment by louis-l** [comment]
Got similar issue when updating from 1.37.0 to 1.38.0

Running on Vercel and MySQL

<img width="1228" alt="image" src="https://user-images.githubusercontent.com/10172565/188805903-638786db-99da-478d-8e16-0de1b90e4a40.png">

**Reply by LoulouCrypto** [reply]
Hi, i have the same problem but i'm runing it with docker .... how can i solved it ?

**Reply by LoulouCrypto** [reply]
2022-09-08 17:46:08.273 UTC [29] LOG:  could not receive data from client: Connection reset by peer
umami-umami-1  | Prisma schema loaded from prisma/schema.prisma
umami-umami-1  | Datasource "db": PostgreSQL database "umami", schema "public" at "db:5432"
umami-umami-1  |
umami-umami-1  | 3 migrations found in prisma/migrations
umami-umami-1  |
umami-umami-1  | Following migration have not yet been applied:
umami-umami-1  | 03_remove_casade_delete
umami-umami-1  |
umami-umami-1  | To apply migrations in development run yarn prisma migrate dev.
umami-umami-1  | To apply migrations in production run yarn prisma migrate deploy.
umami-umami-1  |
umami-umami-1  | Running update...
umami-umami-1  | Error: P3009
umami-umami-1  |
umami-umami-1  | migrate found failed migrations in the target database, new migrations will not be applied. Read more about how to resolve migration issues in a production database: https://pris.ly/d/migrate-resolve
umami-umami-1  | The `02_add_event_data` migration started at 2022-08-10 19:44:26.576604 UTC failed
umami-umami-1  |
umami-umami-1  |
umami-umami-1  | ✗ Command failed: prisma migrate deploy
umami-umami-1  | Error: P3009
umami-umami-1  |
umami-umami-1  | migrate found failed migrations in the target database, new migrations will not be applied. Read more about how to resolve migration issues in a production database: https://pris.ly/d/migrate-resolve
umami-umami-1  | The `02_add_event_data` migration started at 2022-08-10 19:44:26.576604 UTC failed
umami-umami-1  |
umami-umami-1  |
umami-umami-1  |
umami-umami-1  | error Command failed with exit code 1.
umami-umami-1  | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
umami-umami-1  | ERROR: "check-db" exited with 1.
umami-umami-1  | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
umami-umami-1  | error Command failed with exit code 1.
umami-umami-1 exited with code 1

**Reply by LoulouCrypto** [reply]
Ok find on my own with 1.37.0 discutions

**Reply by louis-l** [reply]
For those who have the same issue as mine, just run SSH to mysql server and run `mysql_upgrade` command to fix the "mysql.proc"

Then manually run the missing migrations, then:

```shell
yarn prisma migrate resolve --applied "02_add_event_data"

yarn prisma migrate resolve --applied "03_remove_cascade_delete"
```

**Reply by yuxxeun** [reply]
Thanks, this one was help me, i was rustasing about prisma migration at this point, tbh.

**Comment by pqvst** [comment]
I just upgraded 3 different installations of Umami to version 1.38.0. All of them failed due to the `02_add_event_data` migration.

My installations are running with Docker. I'm not sure if this is the correct way to resolve it, but I did the following:

1) `docker exec -it <container-name> sh`
2) `yarn prisma migrate resolve --applied "02_add_event_data"`

Note: make sure you run it just after the umami container has restarted, otherwise you might not have time to run the yarn command before the container restarts.

**Comment by pqvst** [comment]
Unfortunately my umami instances are still broken. The real-time page is not loading, just showing a blank page. Someone mentioned running "yarn run prisma migrate dev", but I get the following:

```
- The migration `02_add_event_data` was modified after it was applied.
- Drift detected: Your database schema is not in sync with your migration history.

The following is a summary of the differences between the expected database schema given your migrations files, and the actual schema of the database.

It should be understood as the set of changes to get from the expected schema to the actual schema.

[-] Removed tables
  - _event_old
  - event_data

[*] Changed the `event` table
  [-] Removed column `event_name`
  [+] Added column `event_type`
  [+] Added column `event_value`

[*] Changed the `event_data` table
  [-] Removed foreign key on columns (event_id)


✔ We need to reset the PostgreSQL database "umami" at "db:5432".
Do you want to continue? All data will be lost. … no

Reset cancelled.
```

I *really* don't want to reset my data. Any advice?

**Reply by briancao** [reply]
Prisma Migrate Dev, is very opinionated in how migrations are run, and is developer specific.  Try `prisma migrate deploy` or better yet, `yarn check-db` to deploy your migrations.

You can check the 1.37.0 troubleshooting guide to fix errors in deploying your 02_migration.

https://github.com/umami-software/umami/discussions/1434

**Comment by otahirs** [comment]
> We have also made major database changes to get ready for our hosted cloud offering, adding support for Clickhouse, Kafka and Redis. Any changes we make will of course be open-sourced so you will get the same performance benefits we gain

Hi, are there still plans to introduce Clickhouse to the open-sourced version?
